Documentation/SimulatorHID.md

```diff
@@ -0,0 +1,86 @@
+
+## The Indigo Recipient
+
+The `Simulator.app` sending Indigo Messages in Mach explains how HID events in the Mac `Simulator.app` are sent. For this to make meaningful interactions in the Simulator Runtime, there must be a recipient. In the `SimulatorHIDFallback.framework` there is a section that looks a little like:
+
+```
+mach_port_t port port;
+if (bootstrap_look_up(bootstrap_port(), "IndigoHIDRegistrationPort", &port) != KERN_SUCCESS) {
+  // Handle Failure
+}
+// Do something with the port
+```
+
+The `bootstrap` functions are a part of `launchd` that is used to register connections of `mach_port_t`'s without processes having to be directly aware of each other. It can help to think of this as a client-server relationship with the client being the `Simulator.app` and the server  being `SimulatorHIDFallback` running in `backboardd`.
+
+With this relationship established, `SimulatorHIDFallback` can listen for incoming Indigo Messages with the `mach_msg` call. This can be seen in the dissasembly for a block in `-[SimulatorHIDFallbackSystem initUsingHIDService:error:]`. The code would look something like this at a high level:
+
+```
+// Make an IndigoMessage, this will be populated by mach_msg
+IndigoMessage message;
+// mach_msg will block, waiting for an incoming message and the call will return when one is available.
+mach_msg(&message, sizeof(message), 0x0, port, 0x0, 0x0);
+```
+
+This data now has to be interpreted and sent to the underlying HID event system within the Simulator Runtime. Events are sent using the `IOKit.framework`. Parts of this API are public in macOS and there are also a number of Private Components that `SimulatorHIDFallback` uses. This can be seen further down the block:
+
+```
+// The IndigoMessage has a mach_msg_header at the top, we need to get to the contents of the message below it.
+IndigoMessageContents contents = message.contents; // This is message+0x20, the size of the mach_msg_header_t
+// There's always at least one message.
+IOHIDEvent event = __IOHIDEventFromIndigoHIDData(contents, 0x2, 0x0);
+// There can be more HID Events sent in a Single Indigo Message, so repeat on the array of events
+while (event->hasNext) {
+  contents++;
+  IOHIDEvent nextEvent = __IOHIDEventFromIndigoHIDData(contents, 0x2, 0x0);
+  // Update the original event by appending.
+  IOHIDEventAppendEvent(event, nextEvent);
+}
+// Dispatch the event to the Event Service.
+IOHIDServiceDispatchEvent(service, event);
+```
+
+The `__IOHIDEventFromIndigoHIDData` function is clearly quite important in tranforming `Indigo` messages into `IOHIDEvent` instances. The implementation of this function is also quite helpful in re-constructing an idea of what an Indigo message looks like. Inside the implementation of this transformation function there are number of conditionals that change the kind of `IOHIDEvent` (Keyboard/Touch) depending on the data passed in.
```

Comment:
These guys are very on top of their game. It's nice to see all of this written out.